Improve Packages and imports (#5629)
JetBrains/kotlin-web-site: docs/topics/packages.md

[//]: # (title: Packages and imports)

In a Kotlin project, code is organized using packages and imports:

* A **package** is a container for one or more Kotlin files. Files are linked to a package using a `package` header.
* An **import** is a directive that makes entities from other packages available in the current file.

## Package declarations

A source file may start with a package declaration:

```kotlin
package org.example

fun printMessage() { /*...*/ }
class Message(val text: String) { /*...*/ }
```

All contents of the source file (such as classes and functions) belong to this package.
In the example above:

> * The fully qualified name of `printMessage()` is `org.example.printMessage`.

Reviewer: It would be nice to prefix this with a sentence like "The combination of the package name and the name of the declaration is called its _fully qualified name_"

* The fully qualified name of `Message` is `org.example.Message`.

If no package is specified, the file's contents belong to the default root package.

## Imports

To use an entity from a file in a different package, use an `import` directive.
Apart from the default imports, each file may declare its own imports.

> ### Import a single name

Reviewer: ```suggestion
### Import a single declaration
```

Import one specific declaration so it can be used without qualification:

```kotlin
import org.example.Message // Message is accessible without qualification

fun main() {
    val message = Message("Hello")
    println(message.text)
}
```

### Import the contents of a scope

> You can import everything accessible within a scope — a package, class, or object:

Reviewer: This is not correct, you cannot use `*` with an object

[//]: # (title: Packages and imports)

In a Kotlin project, code is organized using packages and imports:

* A **package** is a container for one or more Kotlin files. Files are linked to a package using a `package` header.

> * An **import** is a directive that makes entities from other packages available in the current file.

Reviewer 2: Would this part work with declarations instead of entities here? 🤔 

* An **import** is a directive that makes declarations from other packages available in the current file.

Author: So, I double-checked the Kotlin language specification:
The spec uses "entity" to describe what you are importing and interacting with, while "declaration" is mostly used to describe the syntactic construct or scope.

## Package declarations

A source file may start with a package declaration:

```kotlin
package org.example

fun printMessage() { /*...*/ }
class Message(val text: String) { /*...*/ }
```

> All contents of the source file (such as classes and functions) belong to this package.

Reviewer 2: ```suggestion
All contents of the source file, such as classes and functions, belong to this package.
```

Reviewer 2: (suggestion)
It's mostly preference, but I think you made the sentence read smoothly, so I think it parses well even with commas instead of parentheses.

A declaration's fully qualified name combines the package name with the name of the declaration.

> In the example above:

Reviewer 2: (suggestion)
WDYT about simply: In this example? 🤔 — we are in one paragraph so should be just as understandable and one fewer word (tbh, both works so whichever you prefer)

* The fully qualified name of `printMessage()` is `org.example.printMessage`.
* The fully qualified name of `Message` is `org.example.Message`.

> If no package is specified, the file's contents belong to the default root package.

Reviewer 2: ```suggestion
If a file has no `package` header, its declarations belong to the default root package.
```

Reviewer 2: Just as a nod to the terminology we used at the beginning of the doc, and to avoid the passive structure we could make file the subject.

I'm not sure about the declaration vs contents switch. I'd assume comments would also belong to the root package, but perhaps the focus should be on the declarations, so that's why I suggested it 🤔 
I'm not sure, to me both feel correct, so whichever you think is right 👍

Author: There’s also a switch from “header” to “declaration”! The spec consistently uses "package header"

## Imports

> To use an entity from a file in a different package, use an `import` directive.

Reviewer 2: same here as above, entity vs declaration? 🤔

> Apart from the default imports, each file may declare its own imports.

Reviewer 2: ```suggestion
In addition to the default imports, each file may declare its own imports.
```

### Import a single declaration

> Import one specific declaration so it can be used without qualification:

Reviewer 2: ```suggestion
Import a specific declaration so you can use it without qualification:
```

```kotlin

> import org.example.Message // Message is accessible without qualification

Reviewer 2: In general it may be better to have the comments above (it can avoid scrolling especially for the later ones 🤔 — here I think is fine, but then it might become inconsistent.

fun main() {
    val message = Message("Hello")
    println(message.text)
}
```

### Import the contents of a scope

> Star-imports ending in an asterisk (`*`) import all named entities inside the corresponding scope:

Reviewer 2: ```suggestion
Star imports ending in an asterisk (`*`) import all named declarations inside the corresponding scope:
```

```kotlin

> import org.example.* // Everything in 'org.example' is accessible

Reviewer 2: ```suggestion
import org.example.* // Everything in org.example is accessible
```

fun main() {
    printMessage()
    val message = Message("Hi")
}
```

> If a name is imported via both a star import and an explicit single import,

Reviewer 2: ```suggestion
If you import the same name with both a star import and an explicit single import,
```

Reviewer 2: just to remove latin + passive 🤔 WDYT?

the explicit single import takes priority during name resolution.

### Resolve name clashes with aliases

If two imported names clash, use the `as` keyword to locally rename one of them:

```kotlin

> import org.example.Message             // Message refers to 'org.example.Message'

Reviewer 2: Here, I suggest placing the code comments above to eliminate scrolling and adding a line between them for clarity

    val b = TestMessage("from test")
}
```

### What you can import

The `import` keyword is not limited to classes. You can import any of the following declarations,
whether they come from a package, a class, an object, or an enum:

* Top-level functions and properties declared directly inside a package:
    ```kotlin
    import org.example.printMessage // Top-level function
    import org.example.VERSION      // Top-level property
    ```
* Functions and properties from [object declarations](object-declarations.md#object-declarations-overview):
    ```kotlin
    import org.example.Config.DEFAULT_TIMEOUT // Property from an object
    import org.example.Config.loadSettings    // Function from an object
    ```
* Members of a [companion object](object-declarations.md#companion-objects), referenced through the enclosing class name:
    ```kotlin
    import org.example.MyClass.create // Refers to MyClass.Companion.create
    ```
* [Enum constants](enum-classes.md):
    ```kotlin
    import org.example.Color.RED
    import org.example.Color.GREEN
    ```
* Nested classes:
    ```kotlin
    import org.example.Outer.Nested
    ```

## Default imports

> Kotlin imports several packages into every file by default:

Reviewer: Nitpick: if you say you import a package, you should just write `kotlin.annotations`. Otherwise it's better to say "Kotlin includes the following imports by default" or something like that

* [kotlin.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/index.html)
* [kotlin.annotation.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.annotation/index.html)
* [kotlin.collections.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/index.html)
* [kotlin.comparisons.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.comparisons/index.html)
* [kotlin.io.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.io/index.html)
* [kotlin.ranges.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.ranges/index.html)
* [kotlin.sequences.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.sequences/index.html)
* [kotlin.text.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.text/index.html)
* [kotlin.math.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.math/index.html)

> Additional packages are imported based on the target platform:

Reviewer 2: ```suggestion
Kotlin imports additional packages by default depending on the target platform:
```

Reviewer 2: (suggestion)
again, it's understandable with the passive, but wanted to leave it here as an option

* JVM:
  * [java.lang.*](https://docs.oracle.com/javase/8/docs/api/java/lang/package-summary.html)
  * [kotlin.jvm.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.jvm/index.html)

* JS:
  * [kotlin.js.*](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.js/index.html)

## Visibility and imports

The ability to import an entity depends on its [visibility modifiers](visibility-modifiers.md):

> * `public` entities can be imported anywhere.

Reviewer 2: same question here about declarations vs entities

### Import the contents of a scope

Star imports, ending in an asterisk `*`, import all named entities inside the corresponding scope:

```kotlin
// Everything in `org.example` is accessible
import org.example.* 

fun main() {
    printMessage()
    val message = Message("Hi")
}
```

If you import an entity with both a star import and an explicit import,
the explicit import takes priority during overload resolution.

### Resolve name clashes with aliases

If two imported entities have the same name, use the `as` keyword to locally rename one of them:

```kotlin
// Message refers to `org.example.Message`
import org.example.Message

> // TestMessage refers to `org.test.Message`

Reviewer 2: Just as a last question as a general comment: 
Do you think we need the backticks for these fully qualified names like in: `org.test.Message`? 🤔 I'm asking because for TestMessage we don't add it but for the fully qualified names we do 🤔

Author: Thanks for pointing that out! According to our guidelines, there should be no backticks, so I'll get this fixed right away
